README.md
aio_marantz_avr


HISTORY.rst
=======
History
=======

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog`_, and this project adheres to `Semantic Versioning`_.

.. _`Keep a Changelog`: https://keepachangelog.com/en/1.0.0/
.. _`Semantic Versioning`: https://semver.org/spec/v2.0.0.html

0.1.0 (2020-01-07)
------------------

First release on PyPI.

Added
~~~~~

* Basic control of limited Marantz AVR models.


tests/__init__.py
"""Unit test package for aio_marantz_avr."""


tests/test_avr.py
#!/usr/bin/env python

"""Tests for `avr` package."""

import asyncio
import pytest
import telnetlib3

from typing import Awaitable, List, Mapping

from aio_marantz_avr import (connect, DisconnectedError, InputSource, Power, SurroundMode,
                             AvrTimeoutError)


class TestShell:
    reader = None
    writer = None

    @property
    def connected(self):
        return self.reader is not None and self.writer is not None

    def shell(self, reader, writer):
        self.reader = reader
        self.writer = writer

    async def run_and_respond(self, aw: Awaitable,
                              response_mapping: Mapping[str, List[str]]) -> None:
        await asyncio.wait_for(asyncio.gather(aw, self.expect_and_respond(response_mapping)),
                               timeout=5)

    async def expect_and_respond(self, response_mapping: Mapping[str, List[str]]) -> None:
        while True:
            command = await self.reader.readline()

            if command in response_mapping:
                responses = response_mapping[command]
                del response_mapping[command]
                for response in responses:
                    self.writer.write(response)
                    await self.writer.drain()
                if not response_mapping:
                    return


@pytest.fixture
def test_shell():
    return TestShell()


def create_task(coro):
    # asyncio.create_task does not exist yet in python 3.6, but ensure_future does the same
    return asyncio.ensure_future(coro)


@pytest.fixture
async def test_server(test_shell, unused_tcp_port):
    server = await telnetlib3.create_server(shell=test_shell.shell, port=unused_tcp_port,
                                            encoding="ascii")
    server_task = create_task(server.wait_closed())
    yield server
    test_shell.writer.close()
    server.close()
    await server_task


@pytest.mark.asyncio
async def test_connect(test_server, test_shell, unused_tcp_port):
    avr = await connect("localhost", port=unused_tcp_port)
    assert avr is not None
    assert test_shell.connected


@pytest.fixture
async def avr(test_server, unused_tcp_port):
    return await connect("localhost", port=unused_tcp_port)


@pytest.mark.asyncio
async def test_mute_on(avr, test_shell):
    await test_shell.run_and_respond(
        avr.mute_volume(True),
        {"MUON\r": ["MUON\r"]}
    )
    assert avr.is_volume_muted is True


@pytest.mark.asyncio
async def test_mute_off(avr, test_shell):
    await test_shell.run_and_respond(
        avr.mute_volume(False),
        {"MUOFF\r": ["MUOFF\r"]}
    )
    assert avr.is_volume_muted is False


@pytest.mark.asyncio
async def test_turn_on(avr, test_shell):
    await test_shell.run_and_respond(
        avr.turn_on(),
        {"PWON\r": ["PWON\r"]}
    )
    assert avr.power == Power.On


@pytest.mark.asyncio
async def test_turn_off(avr, test_shell):
    await test_shell.run_and_respond(
        avr.turn_off(),
        {"PWSTANDBY\r": ["PWSTANDBY\r"]}
    )
    assert avr.power == Power.Standby


@pytest.mark.asyncio
async def test_select_source(avr, test_shell):
    await test_shell.run_and_respond(
        avr.select_source(InputSource.DVD),
        {"SIDVD\r": ["SIDVD\r"]}
    )
    assert avr.source == InputSource.DVD


@pytest.mark.asyncio
async def test_select_sound_mode(avr, test_shell):
    await test_shell.run_and_respond(
        avr.select_sound_mode(SurroundMode.Movie),
        {"MSMOVIE\r": ["MSMOVIE\r"]}
    )
    assert avr.sound_mode == SurroundMode.Movie


@pytest.mark.asyncio
async def test_set_volume_level(avr, test_shell):
    await test_shell.run_and_respond(
        avr.set_volume_level(30),
        {"MV30\r": ["MV30\r"]}
    )
    assert avr.volume_level == 30.


@pytest.mark.asyncio
async def test_set_volume_level_single_digit(avr, test_shell):
    await test_shell.run_and_respond(
        avr.set_volume_level(3),
        {"MV03\r": ["MV03\r"]}
    )
    assert avr.volume_level == 3.


@pytest.mark.asyncio
async def test_volume_level_up(avr, test_shell):
    await test_shell.run_and_respond(
        avr.volume_level_up(),
        {"MVUP\r": ["MV30\r"]}
    )
    assert avr.volume_level == 30.


@pytest.mark.asyncio
async def test_volume_level_down(avr, test_shell):
    await test_shell.run_and_respond(
        avr.volume_level_down(),
        {"MVDOWN\r": ["MV30\r"]}
    )
    assert avr.volume_level == 30.


@pytest.mark.asyncio
async def test_volume_level_decimal(avr, test_shell):
    await test_shell.run_and_respond(
        avr.volume_level_down(),
        {"MVDOWN\r": ["MV305\r"]}
    )
    assert avr.volume_level == 30.5


@pytest.mark.asyncio
async def test_refresh(avr, test_shell):
    await test_shell.run_and_respond(
        avr.refresh(),
        {"MU?\r": ["MUOFF\r"], "PW?\r": ["PWON\r"], "MV?\r": ["MV400\r", "MVMAX980\r"],
         "SI?\r": ["SISAT/CBL\r"], "MS?\r": ["MSDOLBY DIGITAL\r"], }
    )
    assert avr.power == Power.On
    assert avr.is_volume_muted is False
    assert avr.volume_level == 40.
    assert avr.max_volume_level == 98.
    assert avr.source == InputSource.CblSat
    assert avr.sound_mode == SurroundMode.DolbyDigital


@pytest.mark.asyncio
async def test_timeout_during_refresh(avr, test_shell):
    task = create_task(test_shell.expect_and_respond({"MU?\r": ["MUOFF\r"], }))

    with pytest.raises(AvrTimeoutError):
        await avr.refresh()

    task.cancel()


def test_initial_state(avr):
    """Initial states are unknown, so None"""
    assert avr.power is None
    assert avr.is_volume_muted is None
    assert avr.volume_level is None
    assert avr.max_volume_level is None
    assert avr.source is None
    assert avr.sound_mode is None


@pytest.mark.asyncio
async def test_refresh_while_running_command(avr, test_shell):
    turn_on_task = create_task(avr.turn_on())
    command = await test_shell.reader.readline()
    assert command == "PWON\r"

    await avr.refresh()

    test_shell.writer.write("PWON\r")
    await test_shell.writer.drain()
    await turn_on_task


@pytest.mark.asyncio
async def test_run_command_while_refreshing(avr, test_shell):
    refresh_task = create_task(avr.refresh())
    command = await test_shell.reader.readline()
    assert command

    await avr.turn_on()
    command2 = await test_shell.reader.readline()
    assert command2 == "PWON\r"

    refresh_task.cancel()


@pytest.mark.asyncio
async def test_refresh_after_disconnect(avr, test_shell, test_server):
    test_shell.writer.close()
    test_server.close()

    with pytest.raises(DisconnectedError):
        await avr.refresh()


@pytest.mark.asyncio
async def test_send_command_after_disconnect(avr, test_shell, test_server):
    test_shell.writer.close()
    test_server.close()

    with pytest.raises(DisconnectedError):
        await avr.turn_on()


@pytest.mark.asyncio
async def test_timeout_in_command(avr):
    with pytest.raises(AvrTimeoutError):
        await avr.turn_on()


tests/test_aio_marantz_avr.py
#!/usr/bin/env python

"""Tests for `aio_marantz_avr` package."""

import asyncio
import pytest
import telnetlib3

from typing import Awaitable, List, Mapping, Optional

from aio_marantz_avr import aio_marantz_avr


class TestShell:
    reader = None
    writer = None

    @property
    def connected(self):
        return self.reader is not None and self.writer is not None

    def shell(self, reader, writer):
        self.reader = reader
        self.writer = writer

    async def run_and_respond(self, aw: Awaitable,
                              response_mapping: Mapping[str, List[str]]) -> None:
        await asyncio.wait_for(asyncio.gather(aw, self.expect_and_respond(response_mapping)),
                               timeout=5)

    async def expect_and_respond(self, response_mapping: Mapping[str, List[str]]) -> None:
        while True:
            command = await self.reader.readline()

            if command in response_mapping:
                responses = response_mapping[command]
                del response_mapping[command]
                for response in responses:
                    self.writer.write(response)
                    await self.writer.drain()
                if not response_mapping:
                    return


@pytest.fixture
def test_shell():
    return TestShell()


@pytest.fixture
async def test_server(test_shell, unused_tcp_port):
    server = await telnetlib3.create_server(shell=test_shell.shell, port=unused_tcp_port,
                                            encoding="ascii")
    server_task = asyncio.create_task(server.wait_closed())
    yield server
    test_shell.writer.close()
    server.close()
    await server_task


@pytest.mark.asyncio
async def test_connect(test_server, test_shell, unused_tcp_port):
    avr = await aio_marantz_avr.connect("localhost", port=unused_tcp_port)
    assert avr is not None
    assert test_shell.connected


@pytest.fixture
async def avr(test_server, unused_tcp_port):
    return await aio_marantz_avr.connect("localhost", port=unused_tcp_port)


@pytest.mark.asyncio
async def test_mute_on(avr, test_shell):
    await test_shell.run_and_respond(
        avr.mute_volume(True),
        {"MUON\r": ["MUON\r"]}
    )
    assert avr.is_volume_muted is True


@pytest.mark.asyncio
async def test_mute_off(avr, test_shell):
    await test_shell.run_and_respond(
        avr.mute_volume(False),
        {"MUOFF\r": ["MUOFF\r"]}
    )
    assert avr.is_volume_muted is False


@pytest.mark.asyncio
async def test_turn_on(avr, test_shell):
    await test_shell.run_and_respond(
        avr.turn_on(),
        {"PWON\r": ["PWON\r"]}
    )
    assert avr.power == aio_marantz_avr.Power.On


@pytest.mark.asyncio
async def test_turn_off(avr, test_shell):
    await test_shell.run_and_respond(
        avr.turn_off(),
        {"PWOFF\r": ["PWOFF\r"]}
    )
    assert avr.power == aio_marantz_avr.Power.Off


@pytest.mark.asyncio
async def test_turn_off_to_standby(avr, test_shell):
    await test_shell.run_and_respond(
        avr.turn_off(),
        {"PWOFF\r": ["PWSTANDBY\r"]}
    )
    assert avr.power == aio_marantz_avr.Power.Standby


@pytest.mark.asyncio
async def test_select_source(avr, test_shell):
    await test_shell.run_and_respond(
        avr.select_source(aio_marantz_avr.InputSource.DVD),
        {"SIDVD\r": ["SIDVD\r"]}
    )
    assert avr.source == aio_marantz_avr.InputSource.DVD


@pytest.mark.asyncio
async def test_select_sound_mode(avr, test_shell):
    await test_shell.run_and_respond(
        avr.select_sound_mode(aio_marantz_avr.SurroundMode.Movie),
        {"MSMOVIE\r": ["MSMOVIE\r"]}
    )
    assert avr.sound_mode == aio_marantz_avr.SurroundMode.Movie


@pytest.mark.asyncio
async def test_refresh(avr, test_shell):
    await test_shell.run_and_respond(
        avr.refresh(),
        {"MU?\r": ["MUOFF\r"], "PW?\r": ["PWON\r"], "MV?\r": ["MV400\r", "MVMAX980\r"],
         "SI?\r": ["SISAT/CBL\r"], "MS?\r": ["MSDOLBY DIGITAL\r"],}
    )
    assert avr.power == aio_marantz_avr.Power.On
    assert avr.is_volume_muted is False
    assert avr.volume_level == 400
    assert avr.max_volume_level == 980
    assert avr.source == aio_marantz_avr.InputSource.CblSat
    assert avr.sound_mode == aio_marantz_avr.SurroundMode.DolbyDigital


def test_initial_state(avr):
    """Initial states are unknown, so None"""
    assert avr.power is None
    assert avr.is_volume_muted is None
    assert avr.volume_level is None
    assert avr.max_volume_level is None
    assert avr.source is None
    assert avr.sound_mode is None


@pytest.mark.asyncio
async def test_refresh_while_running_command(avr, test_shell):
    turn_on_task = asyncio.create_task(avr.turn_on())
    command = await test_shell.reader.readline()
    assert command == "PWON\r"

    await avr.refresh()

    test_shell.writer.write("PWON\r")
    await test_shell.writer.drain()
    await turn_on_task


@pytest.mark.asyncio
async def test_run_command_while_refreshing(avr, test_shell):
    refresh_task = asyncio.create_task(avr.refresh())
    command = await test_shell.reader.readline()
    assert command

    await avr.turn_on()
    command2 = await test_shell.reader.readline()
    assert command2 == "PWON\r"

    refresh_task.cancel()



aio_marantz_avr/avr.py
"""Control of an AVR over Telnet."""

import asyncio
import telnetlib3

from typing import Any, List, MutableMapping, Optional, Type

from .enums import InputSource, Power, SurroundMode


class AvrError(Exception):
    """Base class for all errors returned from an AVR."""

    pass


class DisconnectedError(AvrError):
    """The connection to the AVR has been lost, or has never been established."""

    pass


class AvrTimeoutError(AvrError):
    """A request to the AVR has timed out."""

    pass


async def connect(host: str, port: int = 23, timeout: float = 1.0) -> "MarantzAVR":
    """Connect to an AVR."""
    reader, writer = await telnetlib3.open_connection(host, port=port, encoding="ascii")
    return MarantzAVR(reader, writer, timeout)


def _on_off_from_bool(value: bool) -> str:
    if value:
        return "ON"
    else:
        return "OFF"


def _on_off_to_bool(value: str) -> bool:
    return value == "ON"


class _DataDefinition:
    query: str
    data_names: List[str]

    def __init__(self, query: str, data_names: List[str]):
        self.query = query
        self.data_names = data_names


class MarantzAVR:
    """Connection to a Marantz AVR over Telnet.

    Uses `connect` to create a connection to the AVR.
    """

    DATA_DEFS: List[_DataDefinition] = [
        _DataDefinition("PW?", ["PW"]),
        _DataDefinition("MU?", ["MU"]),
        _DataDefinition("MV?", ["MV", "MVMAX"]),
        _DataDefinition("SI?", ["SI"]),
        _DataDefinition("MS?", ["MS"]),
    ]

    _reader: telnetlib3.TelnetReader
    _writer: telnetlib3.TelnetWriter
    _timeout: float

    _data: MutableMapping[str, Optional[str]]
    _reading: bool

    def __init__(
        self, reader: telnetlib3.TelnetReader, writer: telnetlib3.TelnetWriter, timeout: float,
    ):
        self._reader = reader
        self._writer = writer
        self._timeout = timeout
        self._prepare_data()
        self._reading = False

    def _prepare_data(self) -> None:
        self._data = {}
        for data_def in self.DATA_DEFS:
            for name in data_def.data_names:
                self._data[name] = None

    @property
    def power(self) -> Optional[Power]:
        """Power state of the AVR."""
        return self._get_enum_data("PW", Power)

    @property
    def is_volume_muted(self) -> Optional[bool]:
        """Boolean if volume is currently muted."""
        return self._get_boolean_data("MU")

    @property
    def volume_level(self) -> Optional[float]:
        """Volume level of the AVR zone (00..max_volume_level)."""
        return self._get_volume_level("MV")

    @property
    def max_volume_level(self) -> Optional[float]:
        """Maximum volume level of the AVR zone."""
        return self._get_volume_level("MVMAX")

    @property
    def source(self) -> Optional[InputSource]:
        """Name of the current input source."""
        return self._get_enum_data("SI", InputSource)

    @property
    def source_list(self) -> List[InputSource]:
        """List of available input sources."""
        return list(InputSource)

    @property
    def sound_mode(self) -> Optional[SurroundMode]:
        """Name of the current sound mode."""
        return self._get_enum_data("MS", SurroundMode)

    @property
    def sound_mode_list(self) -> List[SurroundMode]:
        """List of available sound modes."""
        return list(SurroundMode)

    async def refresh(self) -> None:
        """Refresh all properties from the AVR."""
        if self._reading:
            return

        for data_def in self.DATA_DEFS:
            await self._send_command(data_def.query)
            await self._wait_for_response_with_timeout(*data_def.data_names)

    async def turn_on(self) -> None:
        """Turn the AVR on."""
        await self._send_command("PW", "ON")
        await self._wait_for_response_with_timeout("PW")

    async def turn_off(self) -> None:
        """Turn the AVR off."""
        await self._send_command("PW", "STANDBY")
        await self._wait_for_response_with_timeout("PW")

    async def mute_volume(self, mute: bool) -> None:
        """Mute or unmute the volume.

        Arguments:
        mute -- True to mute, False to unmute.
        """
        await self._send_command("MU", _on_off_from_bool(mute))
        await self._wait_for_response_with_timeout("MU")

    async def set_volume_level(self, level: int) -> None:
        """Set the volume level.

        Arguments:
        level -- An integer value between 0 and `max_volume_level`.
        """
        await self._send_command(f"MV{level:02}")
        await self._wait_for_response_with_timeout("MV")

    async def volume_level_up(self) -> None:
        """Turn the volume level up one notch."""
        await self._send_command("MVUP")
        await self._wait_for_response_with_timeout("MV")

    async def volume_level_down(self) -> None:
        """Turn the volume level down one notch."""
        await self._send_command("MVDOWN")
        await self._wait_for_response_with_timeout("MV")

    async def select_source(self, source: InputSource) -> None:
        """Select the input source."""
        await self._send_command("SI", source.value)
        await self._wait_for_response_with_timeout("SI")

    async def select_sound_mode(self, mode: SurroundMode) -> None:
        """Select the sound mode."""
        await self._send_command("MS", mode.value)
        await self._wait_for_response_with_timeout("MS")

    async def _send_command(self, *parts: str) -> None:
        if self._reader.at_eof():
            raise DisconnectedError()

        self._writer.write("".join(parts))
        self._writer.write("\r")
        await self._writer.drain()

    async def _with_timeout(self, coro) -> Optional[Any]:
        try:
            return await asyncio.wait_for(coro, self._timeout)
        except asyncio.TimeoutError:
            raise AvrTimeoutError

    async def _wait_for_response_with_timeout(self, *names: str) -> None:
        await self._with_timeout(self._wait_for_response(*names))

    async def _wait_for_response(self, *names: str) -> None:
        if self._reading:
            return

        self._reading = True
        try:
            pending_names = list(names)
            while True:
                line = await self._reader.readline()
                if not line and self._reader.at_eof():
                    raise DisconnectedError

                match = self._process_response(line)
                if match in pending_names:
                    pending_names.remove(match)
                    if len(pending_names) == 0:
                        return
        except ConnectionError:
            raise DisconnectedError
        finally:
            self._reading = False

    def _process_response(self, response: str) -> Optional[str]:
        matches = [name for name in self._data.keys() if response.startswith(name)]

        if not matches:
            return None

        if len(matches) > 1:
            matches.sort(key=len, reverse=True)
        match = matches[0]

        self._data[match] = response.strip()[len(match) :]
        return match

    def _get_boolean_data(self, name: str) -> Optional[bool]:
        value = self._data[name]
        if value is not None:
            return _on_off_to_bool(value)
        return None

    def _get_int_data(self, name: str) -> Optional[int]:
        value = self._data[name]
        if value is not None:
            return int(value)
        return None

    def _get_enum_data(self, name: str, enum_type: Type) -> Optional[Any]:
        value = self._data[name]
        if value is not None:
            return enum_type(value)
        return None

    def _get_volume_level(self, name: str) -> Optional[float]:
        value = self._data[name]
        if value is None:
            return None

        # Volume levels are either 2 or 3 characters. The floating point is between 2 and 3.
        value = value.strip()
        level = float(value)
        if len(value) == 3:
            level /= 10.0
        return level


aio_marantz_avr/cli.py
"""Console script for aio_marantz_avr."""
import argparse
import asyncio
import sys

from .avr import connect, DisconnectedError
from .enums import InputSource, SurroundMode


async def run(args):
    avr = await connect(host=args.host)

    if args.show:
        await avr.refresh()
        print(f"Power: {avr.power}")
        print(f"Is volume muted: {avr.is_volume_muted}")
        print(f"Volume level: {avr.volume_level}")
        print(f"Max volume level: {avr.max_volume_level}")
        print(f"Source: {avr.source}")
        print(f"Sound mode: {avr.sound_mode}")

    if args.turn_on:
        await avr.turn_on()
        print(f"NEW Power: {avr.power}")

    if args.turn_off:
        await avr.turn_off()
        print(f"NEW Power: {avr.power}")

    if args.mute_volume_on:
        await avr.mute_volume(True)
        print(f"NEW Is volume muted: {avr.is_volume_muted}")

    if args.mute_volume_off:
        await avr.mute_volume(False)
        print(f"NEW Is volume muted: {avr.is_volume_muted}")

    if args.select_source:
        await avr.select_source(InputSource[args.select_source])
        print(f"NEW Source: {avr.source}")

    if args.select_sound_mode:
        await avr.select_sound_mode(SurroundMode[args.select_sound_mode])
        print(f"NEW Sound mode: {avr.sound_mode}")

    if args.set_volume_level:
        await avr.set_volume_level(args.set_volume_level)
        print(f"NEW Volume level: {avr.volume_level}")

    if args.volume_level_up:
        await avr.volume_level_up()
        print(f"NEW Volume level: {avr.volume_level}")

    if args.volume_level_down:
        await avr.volume_level_down()
        print(f"NEW Volume level: {avr.volume_level}")


def main():
    """Console script for aio_marantz_avr."""
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", required=True, help="Host name or IP of the AVR.")
    parser.add_argument("--show", action="store_true", help="Show all properties.")
    parser.add_argument("--turn-on", action="store_true", help="Turn the AVR on.")
    parser.add_argument("--turn-off", action="store_true", help="Turn the AVR off.")
    parser.add_argument("--mute-volume-on", action="store_true", help="Turn volume mute on.")
    parser.add_argument("--mute-volume-off", action="store_true", help="Turn volume mute off.")
    parser.add_argument(
        "--select-source", choices=InputSource.__members__.keys(), help="Select input source.",
    )
    parser.add_argument(
        "--select-sound-mode", choices=SurroundMode.__members__.keys(), help="Select sound mode.",
    )
    parser.add_argument("--set-volume-level", type=int, help="Set the volume level.")
    parser.add_argument("--volume-level-up", action="store_true", help="Turn the volume level up.")
    parser.add_argument(
        "--volume-level-down", action="store_true", help="Turn the volume level down."
    )
    args = parser.parse_args()

    try:
        asyncio.run(run(args))
    except DisconnectedError:
        print("Connection lost.", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


aio_marantz_avr/enums.py
"""Enums used by the AVR."""

from enum import Enum


class Power(Enum):
    Off = "OFF"
    On = "ON"
    Standby = "STANDBY"


class InputSource(Enum):
    Phono = "PHONO"
    CD = "CD"
    DVD = "DVD"
    Bluray = "BD"
    TV = "TV"
    CblSat = "SAT/CBL"
    MediaPlayer = "MPLAY"
    Game = "GAME"
    Tuner = "TUNER"
    HDRadio = "HDRADIO"
    SiriusXM = "SIRIUSXM"
    Pandora = "PANDORA"
    InternetRadio = "IRADIO"
    Server = "SERVER"
    Favorites = "FAVORITES"
    Aux1 = "AUX1"
    Aux2 = "AUX2"
    Aux3 = "AUX3"
    Aux4 = "AUX4"
    Aux5 = "AUX5"
    Aux6 = "AUX6"
    Aux7 = "AUX7"
    OnlineMusic = "NET"
    Bluetooth = "BT"


class SurroundMode(Enum):
    # Settable values
    Movie = "MOVIE"
    Music = "MUSIC"
    Game = "GAME"
    Direct = "DIRECT"
    PureDirect = "PURE DIRECT"
    Stereo = "STEREO"
    Auto = "AUTO"
    DolbyDigital = "DOLBY DIGITAL"
    DtsSurround = "DTS SURROUND"
    Auro3D = "AURO3D"
    Auro2DSurround = "AURO2DSURR"
    MultiChannelStereo = "MCH STEREO"
    Virtual = "VIRTUAL"

    # Rotate between options
    Left = "LEFT"
    Right = "RIGHT"

    # Return only
    # TODO: Split combined modes
    DolbySurround = "DOLBY SURROUND"
    DolbyAtmos = "DOLBY ATMOS"
    DolbyDigitalDS = "DOLBY D+DS"
    DolbyDigitalNeuralX = "DOLBY D+NEURAL:X"
    DolbyDigitalPlus = "DOLBY D+"
    DolbyDigitalPlusDS = "DOLBY D+ +DS"
    DolbyDigitalPlusNeuralX = "DOLBY D+ +NEURAL:X"
    DolbyHD = "DOLBY HD"
    DolbyHDDS = "DOLBY HD+DS"
    DolbyHDNeuralX = "DOLBY HD+NEURAL:X"
    NeuralX = "NEURAL:X"
    DtsEsDscrt6_1 = "DTS ES DSCRT6.1"
    DtsEsMtrx6_1 = "DTS ES MTRX6.1"
    DtsDS = "DTS+DS"
    DtsNeuralX = "DTS+NEURAL:X"
    DtsEsMtrxNeuralX = "DTS ES MTRX+NEURAL:X"
    DtsEsDscrtNeuralX = "DTS ES DSCRT+NEURAL:X"
    Dts96_24 = "DTS96/24"
    Dts96EsMtrx = "DTS96 ES MTRX"
    DtsHD = "DTS HD"
    DtsHDMstr = "DTS HD MSTR"
    DtsHDDS = "MSDTS HD+DS"
    DtsHDNeuralX = "DTS HD+NEURAL:X"
    DtsX = "DTS:X"
    DtsXMstr = "DTS:X MSTR"
    DtsExpress = "DTS EXPRESS"
    DtsES8ChDscrt = "DTS ES 8CH DSCRT"
    MultiChIn = "MULTI CH IN"
    MultiChInDS = "M CH IN+DS"
    MultiChInNeuralX = "M CH IN+NEURAL:X"
    MultiChIn7_1 = "MULTI CH IN 7.1"


aio_marantz_avr/aio_marantz_avr.py
"""Main module."""

import telnetlib3

from enum import Enum
from typing import Any, List, Mapping, Optional, Tuple, Type


async def connect(host: str, port: int = 23):
    reader, writer = await telnetlib3.open_connection(host, port=port, encoding="ascii")
    return MarantzAVR(reader, writer)


def _on_off_from_bool(value: bool) -> str:
    if value:
        return "ON"
    else:
        return "OFF"


def _on_off_to_bool(value: str) -> bool:
    return value == "ON"


class Power(Enum):
    Off = "OFF"
    On = "ON"
    Standby = "STANDBY"


class InputSource(Enum):
    Phono = "PHONO"
    CD = "CD"
    DVD = "DVD"
    Bluray = "BD"
    TV = "TV"
    CblSat = "SAT/CBL"
    MediaPlayer = "MPLAY"
    Game = "GAME"
    Tuner = "TUNER"
    HDRadio = "HDRADIO"
    SiriusXM = "SIRIUSXM"
    Pandora = "PANDORA"
    InternetRadio = "IRADIO"
    Server = "SERVER"
    Favorites = "FAVORITES"
    Aux1 = "AUX1"
    Aux2 = "AUX2"
    Aux3 = "AUX3"
    Aux4 = "AUX4"
    Aux5 = "AUX5"
    Aux6 = "AUX6"
    Aux7 = "AUX7"
    OnlineMusic = "NET"
    Bluetooth = "BT"


class SurroundMode(Enum):
    # Settable values
    Movie = "MOVIE"
    Music = "MUSIC"
    Game = "GAME"
    Direct = "DIRECT"
    PureDirect = "PURE DIRECT"
    Stereo = "STEREO"
    Auto = "AUTO"
    DolbyDigital = "DOLBY DIGITAL"
    DtsSurround = "DTS SURROUND"
    Auro3D = "AURO3D"
    Auro2DSurround = "AURO2DSURR"
    MultiChannelStereo = "MCH STEREO"
    Virtual = "VIRTUAL"

    # Rotate between options
    Left = "LEFT"
    Right = "RIGHT"

    # Return only
    # TODO: Split combined modes
    DolbySurround = "DOLBY SURROUND"
    DolbyAtmos = "DOLBY ATMOS"
    DolbyDigitalDS = "DOLBY D+DS"
    DolbyDigitalNeuralX = "DOLBY D+NEURAL:X"
    DolbyDigitalPlus = "DOLBY D+"
    DolbyDigitalPlusDS = "DOLBY D+ +DS"
    DolbyDigitalPlusNeuralX = "DOLBY D+ +NEURAL:X"
    DolbyHD = "DOLBY HD"
    DolbyHDDS = "DOLBY HD+DS"
    DolbyHDNeuralX = "DOLBY HD+NEURAL:X"
    NeuralX = "NEURAL:X"
    DtsEsDscrt6_1 = "DTS ES DSCRT6.1"
    DtsEsMtrx6_1 = "DTS ES MTRX6.1"
    DtsDS = "DTS+DS"
    DtsNeuralX = "DTS+NEURAL:X"
    DtsEsMtrxNeuralX = "DTS ES MTRX+NEURAL:X"
    DtsEsDscrtNeuralX = "DTS ES DSCRT+NEURAL:X"
    Dts96_24 = "DTS96/24"
    Dts96EsMtrx = "DTS96 ES MTRX"
    DtsHD = "DTS HD"
    DtsHDMstr = "DTS HD MSTR"
    DtsHDDS = "MSDTS HD+DS"
    DtsHDNeuralX = "DTS HD+NEURAL:X"
    DtsX = "DTS:X"
    DtsXMstr = "DTS:X MSTR"
    DtsExpress = "DTS EXPRESS"
    DtsES8ChDscrt = "DTS ES 8CH DSCRT"
    MultiChIn = "MULTI CH IN"
    MultiChInDS = "M CH IN+DS"
    MultiChInNeuralX = "M CH IN+NEURAL:X"
    MultiChIn7_1 = "MULTI CH IN 7.1"


class DataDefinition:
    query: str
    data_names: List[str]

    def __init__(self, query: str, data_names: List[str]):
        self.query = query
        self.data_names = data_names

class MarantzAVR:
    DATA_DEFS: List[DataDefinition] = [
        DataDefinition("PW?", ["PW"]),
        DataDefinition("MU?", ["MU"]),
        DataDefinition("MV?", ["MV", "MVMAX"]),
        DataDefinition("SI?", ["SI"]),
        DataDefinition("MS?", ["MS"]),
    ]

    _reader: telnetlib3.TelnetReader
    _writer: telnetlib3.TelnetWriter

    _data: Mapping[str, Optional[str]]
    _reading: bool

    def __init__(self, reader: telnetlib3.TelnetReader, writer: telnetlib3.TelnetWriter):
        self._reader = reader
        self._writer = writer
        self._prepare_data()
        self._reading = False

    def _prepare_data(self):
        self._data = {}
        for data_def in self. DATA_DEFS:
            for name in data_def.data_names:
                self._data[name] = None

    @property
    def power(self) -> Optional[Power]:
        """Power state of the AVR."""
        return self._get_enum_data("PW", Power)

    @property
    def is_volume_muted(self) -> Optional[bool]:
        """Boolean if volume is currently muted."""
        return self._get_boolean_data("MU")

    @property
    def volume_level(self) -> Optional[int]:
        """Volume level of the AVR zone (00..max_volume_level)."""
        return self._get_int_data("MV")

    @property
    def max_volume_level(self) -> Optional[int]:
        """Maximum volume level of the AVR zone."""
        return self._get_int_data("MVMAX")

    @property
    def source(self) -> Optional[InputSource]:
        """Name of the current input source."""
        return self._get_enum_data("SI", InputSource)

    @property
    def source_list(self) -> List[InputSource]:
        """List of available input sources."""
        return list(InputSource)

    @property
    def sound_mode(self) -> Optional[SurroundMode]:
        """Name of the current sound mode."""
        return self._get_enum_data("MS", SurroundMode)

    @property
    def sound_mode_list(self) -> List[SurroundMode]:
        """List of available sound modes."""
        return list(SurroundMode)

    async def refresh(self) -> None:
        if self._reading:
            return

        for data_def in self.DATA_DEFS:
            await self._send_command(data_def.query)
            await self._wait_for_response(*data_def.data_names)

    async def turn_on(self) -> None:
        await self._send_command("PW", "ON")
        await self._wait_for_response("PW")

    async def turn_off(self) -> None:
        await self._send_command("PW", "OFF")
        await self._wait_for_response("PW")

    async def mute_volume(self, mute: bool) -> None:
        await self._send_command("MU", _on_off_from_bool(mute))
        await self._wait_for_response("MU")

    async def select_source(self, source: InputSource) -> None:
        await self._send_command("SI", source.value)
        await self._wait_for_response("SI")

    async def select_sound_mode(self, mode: SurroundMode) -> None:
        await self._send_command("MS", mode.value)
        await self._wait_for_response("MS")

    async def _send_command(self, *parts: Tuple[str]) -> None:
        self._writer.write("".join(parts))
        self._writer.write("\r")
        await self._writer.drain()

    async def _wait_for_response(self, *names: Tuple[str]) -> None:
        if self._reading:
            return

        self._reading = True
        try:
            names = list(names)
            while True:
                line = await self._reader.readline()
                match = self._process_response(line)
                if match in names:
                    names.remove(match)
                    if len(names) == 0:
                        return
        finally:
            self._reading = False

    def _process_response(self, response: str) -> Optional[str]:
        matches = [name for name in self._data.keys() if response.startswith(name)]

        if not matches:
            return None

        if len(matches) > 1:
            matches.sort(key=len, reverse=True)
        match = matches[0]

        self._data[match] = response.strip()[len(match):]
        return match

    def _get_boolean_data(self, name: str) -> Optional[bool]:
        value = self._data[name]
        if value is not None:
            return _on_off_to_bool(self._data[name])
        return None

    def _get_int_data(self, name: str) -> Optional[int]:
        value = self._data[name]
        if value is not None:
            return int(value)
        return None

    def _get_enum_data(self, name: str, enum_type: Type) -> Optional[Any]:
        value = self._data[name]
        if value is not None:
            return enum_type(value)
        return None
